### ISSUE TITLE
messages aren't being printed into cider repl

### ISSUE THREAD

**Issue by zck** [clojure-emacs/cider#2138]
I'm seeing the same problem as a [few](https://github.com/clojure-emacs/cider/issues/793) closed [bugs](https://github.com/clojure-emacs/cider/issues/853). The problem was listed as fixed in the second bug report, so I'm opening a new bug.

## Steps to reproduce the problem

Put this in a clojure buffer:

```
(ns quil-art.test-print
  (:require [quil.core :as q]))

(defn print-setup []
  (printf "yep")
  (flush))


(q/defsketch test-print
  :setup print-setup
  :size [100 100])
```

Run `cider-jack-in`

Evaluate each of the lines.

## Expected behavior

"yep" is printed in the cider-repl buffer.

## Actual behavior

Nothing is printed in the buffer.

If I then set `nrepl-log-messages` to t, I then start to get errors like this:

```
error in process filter: nrepl--dispatch-response: [nREPL] No response handler with id nil found
error in process filter: [nREPL] No response handler with id nil found
```

Then, after looking in the *nrepl-messages* buffer, I see messages like this:

```
(<--
  session    "ce243c8f-ca3c-4fe9-a8d1-0946e963ab1b"
  time-stamp "2017-12-13 01:40:05.472154507"
  out        "yep"
)
```

If I merely evaluate `(printf "yep")`, or even `(print-setup)`, I get the messages in the repl as expected. It seems to only happen when the function is passed to the `defsketch` call. (I'm posting here because there have been some previous similar issues in this repo). Let me know if I can do anything to help debug this more. Thanks.

### CIDER version information

```
;; CIDER 0.16.0snapshot (package: 20171211.846), nREPL 0.2.13
;; Clojure 1.8.0, Java 1.8.0_151
```

### Emacs version

Emacs 26.0.50 (built from git HEAD a few months ago).

### Operating system

Ubuntu 16.04

**Comment by gonewest818** [comment]
The suggestion from /u/r_b__ka [on your reddit thread](https://www.reddit.com/r/Clojure/comments/7e6kf6/why_does_quil_the_clojure_library_seem_to_buffer/) works for me. In other words edit your sketch like this

```
(q/defsketch test-print
  :features [:no-bind-output]
  :setup print-setup
  :size [100 100])
```

and reevaluate the `defsketch` form. When I do this I get "yep" printed in the repl (and I also get another window named "Applet" on the desktop). Whenever I restart the repl, I notice the first "yep" prints to the buffer named `*nrepl-server <name>*`.

cider 0.15.1 and nrepl 0.2.13 on emacs 25.3.

**Comment by bbatsov** [comment]
Well, I said before it seems to me this must be some bug in nREPL that we need to investigate and solve. Callback handlers are attached, based on the id of the request and if some responses start coming without a request id we simply don't know what to do. I can add a workaround for this, but I don't like the idea that much.

**Comment by zck** [comment]
With some quick testing, adding `:features [:no-bind-output]` seems to be working. That's weird, because I tried it when seeing it on reddit, and it didn't help then.

I'll give it some more tries this evening, and see if it's completely working or not.

**Comment by zck** [comment]
Ugh, it doesn't seem to be working. This is a different thing from what I'm seeing before.

Instead of getting a message in the `*nrepl-messages*` buffer with no id, I'm not getting that any message back into the `*nrepl-messages*` buffer.

For reference, the complete code I'm testing right now:

```
(ns quil-art.test-print
  (:require [quil.core :as q]))

(defn print-setup []
  (printf "yep\n"))

(q/defsketch test-print
  :features [:no-bind-output]
  :setup print-setup
  :size [100 100])
```

In this example, I get messages up to id 15 in the `*nrepl-messages*` buffer. None of these has the "yep" message coming *back* to the repl.

Then, if I go to the repl buffer, type `3`, and press `enter`, I get this in the repl:

```
user> 3
yep
3
user> 
```

And these messages in the `*nrepl-messages*` buffer:

```
(-->
  id         "16"
  op         "eval"
  session    "9d2707fb-a9a9-4e4e-936c-f0bdc48c7bb6"
  time-stamp "2017-12-14 00:20:07.140569398"
  code       "3
"
  column     7
  file       "*cider-repl quil-art*"
  line       43
  ns         "user"
)
(<--
  id         "16"
  session    "9d2707fb-a9a9-4e4e-936c-f0bdc48c7bb6"
  time-stamp "2017-12-14 00:20:07.158385333"
  out        "yep
"
)
(<--
  id         "16"
  session    "9d2707fb-a9a9-4e4e-936c-f0bdc48c7bb6"
  time-stamp "2017-12-14 00:20:07.161143690"
  ns         "user"
  value      "3"
)
```

Note that the "yep" coming back has a timestamp after I typed `3` in the repl.

I'm happy to refile this bug against another project if that makes more sense. This behavior is certainly frustrating, so I'd like to do what I can to actually fix it.

**Comment by gonewest818** [comment]
I'm not sure how else to help... I'm not getting the same behavior as you but I'm also using different versions of emacs, cider and nrepl. For reference, this code works for me:

```
(ns quiltest.core
  (:require [quil.core :as q]))

(def z (atom true))

(defn setup []
  (println "setup")
  (q/frame-rate 10))

(defn draw []
  (q/background 0 0 0)
  (q/text (if @z "yep" "nope") 20 20))

(defn toggle []
  (swap! z not)
  (println "z is now" @z))

(q/defsketch test-print
  :title "abcdefg"
  :features [:no-bind-output]
  :setup setup
  :draw draw
  :key-pressed toggle
  :size [300 300])
```

When I first `cider-jack-in` the message "setup" lands in the buffer named `*nrepl-server ...*`.  However from that point forward, every subsequent keypress calls the `toggle` callback and prints "z is now {true|false}" to the repl.

**Comment by gonewest818** [comment]
fwiw, the code I posted above also works with cider 0.16.0snapshot and nrepl 0.2.13 on emacs 25.3.  Same behavior as described in the previous comment.

**Comment by zck** [comment]
I'm not really sure what to do next either. When I've tested the code you provide, it works. But I constantly get into cases where it does not.

**Comment by zck** [comment]
I will say that I do appreciate everyone who's looked into it.

**Comment by bbatsov** [comment]
So, we can certainly change CIDER to handle `out` messages without an `id` and print them somewhere, but the real mystery is where exactly is this output originating from, if it's not the result of a request. Likely it's something to do with logging that gets binding to the output stream before nREPL kicks in or something like that. Haven't look into this for a while and don't recall the details of what exactly was going wrong with logging and the server's output. Hopefully we'll be to address this in nREPL itself somehow once @cemerick is ready with 0.3 and the "reborn" nREPL is out in the world.

**Comment by bbatsov** [comment]
And for the sake of completeness - that's where the practical problem occurs https://github.com/clojure-emacs/cider/blob/master/nrepl-client.el#L489

Here we get some `out` response message without a callback handler. We can address this by using some generic `out` handling in such cases, but eventually we have to solve the real problem as well.

**Comment by bbatsov** [comment]
And a bit of historical perspective here https://github.com/clojure-emacs/cider-nrepl/pull/387

**Comment by gonewest818** [comment]
In the meantime depending on your needs you can

(a) log to a file and watch the file (bbatsov mentioned this in one of those tickets somewhere)

(b) display data in the quil sketch as an overlay you can toggle on and off, e.g. https://bugs.mojang.com/secure/attachment/50237/debug-overlay.png

**Comment by NightMachinary** [comment]
Any progress?

**Comment by bbatsov** [comment]
Nope. I've never encountered this problem myself and it seems no one else tried to debug it further.

**Comment by onetom** [comment]
Messages without an `id` can originate from the official `:greeting-fn` parameter of `nrepl.server/start-server`, so it's a very valid use-case, which should be handled somehow.

The only example of using a `:greeting-fn` in nREPL's source is here:
https://github.com/nrepl/nrepl/blob/4eb9ad189f97dc88c70b4163a6c2ca327f9a95fc/doc/modules/ROOT/pages/design/transports.adoc#L148

which showcases the [`nrepl.transport/tty-greeting`](https://github.com/nrepl/nrepl/blob/14d1d959950f692a17dd1197d66f6ef254add1cf/src/clojure/nrepl/transport.clj#L209-L221) function, which just does a `(send transport {:out "<some string>"})`, because it doesn't have access to any session info or any message id, because it's just called like this in `nrepl.server/accept-connection-loop`:
https://github.com/nrepl/nrepl/blob/14d1d959950f692a17dd1197d66f6ef254add1cf/src/clojure/nrepl/server.clj#L84

**Comment by bbatsov** [comment]
@onetom It's best to file a new ticket for this.
